### PR TITLE
Translate runtime-known barrier's MemScope from SPV-IR to OCL2.0 IR

### PR DESCRIPTION
**Pull request by MrSidims**
This partially addresses https://github.com/KhronosGroup/SPIRV-LLVM-Translator/issues/1805

Signed-off-by: Sidorov, Dmitry <dmitry.sidorov@intel.com>

### CODE REVIEW

**Comment by MrSidims** [comment]
I have a WIP patch locally, that fixes forward translation from OCL to SPIR-V, but it's not yet finished. Reverse translation is a bit of more importance for us now, so this part was separated from the main effort to speed up a fix for a bug we see in SYCL.

@svenvh could you please take a look?

File: test/transcoding/barrier-runtime-scope.ll [hunk 1]
```diff
@@ -0,0 +1,50 @@
+; RUN: llvm-as %s -o %t.bc
```
[the remark this answers is not in the document]
Comment (svenvh) [reply test/transcoding/barrier-runtime-scope.ll:1]: Perhaps it's useful to add a TODO comment in this test?

Comment (MrSidims) [inline test/transcoding/barrier-runtime-scope.ll:1]:
```suggestion
; This test checks that the translator is capable to correctly translate
; __spirv_ControlBarrier with runtime-known MemScope parameter
; to SPIR-V and back to OpenCL 2.0 IR.
; TODO: to remove this test once
; https://github.com/KhronosGroup/SPIRV-LLVM-Translator/issues/1805
; is fixed as barrier.cl and sub_group_barrier.cl will be enough to test this
; case

; RUN: llvm-as %s -o %t.bc
```

[the remark this answers is not in the document]
Comment (MrSidims) [reply test/transcoding/barrier-runtime-scope.ll:9]: Added, thanks!

**Review by svenvh — APPROVED** [review]

File: test/transcoding/barrier-runtime-scope.ll [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+; is fixed as barrier.cl and sub_group_barrier.cl will be enough to test this
+; case
+
+; RUN: llvm-as %s -o %t.bc
+; RUN: llvm-spirv %t.bc -spirv-text -o %t.spv.txt
+; RUN: FileCheck < %t.spv.txt %s --check-prefix=CHECK-SPIRV
+; RUN: llvm-spirv -to-binary %t.spv.txt -o %t.spv
+; RUN: spirv-val %t.spv
+; RUN: llvm-spirv --spirv-target-env=CL2.0 -r %t.spv -o %t.rev.bc
+; RUN: llvm-dis %t.rev.bc | FileCheck < %t.rev.ll %s --check-prefix=CHECK-LLVM
```
Comment (svenvh) [inline test/transcoding/barrier-runtime-scope.ll:15]: Actually there's a bug in this line I think? See the CI failures.

Comment (MrSidims) [reply test/transcoding/barrier-runtime-scope.ll:15]: Yes, it should probably be something like that:
llvm-dis %t.bc -o - | FileCheck %s --check-prefix=CHECK-LLVM

**Pull request merged**